Initial support for multiple congregations

ajdlinux opened the pull request:
Add a new Congregation model and allow users to choose a congregation when adding a new service.

Currently, we don't provide any way to query congregation information once it's in the database, or otherwise display it in any of the views, but this is a start.

Signed-off-by: Andrew Donnellan <andrew@donnellan.id.au>

ajdlinux committed:
Create Congregation model

Add a basic model to represent multiple congregations.

Signed-off-by: Andrew Donnellan <andrew@donnellan.id.au>

ajdlinux committed:
service: add congregation field when creating new service

Allow the user to specify the congregation when creating a new service.

Signed-off-by: Andrew Donnellan <andrew@donnellan.id.au>

tessereth commented on music/app/controllers/services_controller.rb:
```diff
@@ -11,7 +11,9 @@ def new
     @resource = OpenStruct.new(date: Date.parse(params[:resource][:date]), body: params[:resource][:body])
     begin
-      s = Service.from_email(params[:resource][:date], params[:resource][:body])
+      congregation = Congregation.find_by id: params[:resource][:congregation_id]
```
While you don't need brackets around parameters to functions, best practice says you should add them anyway.

ajdlinux replied:
Ruby is weird

tessereth commented on music/app/views/services/new.html.haml:
```diff
@@ -12,6 +12,8 @@
 %h2 Enter next service
 = form_for :resource, url: new_service_path do |f|
+  %p= f.label :congregation_id, 'Congregation'
+  %p= f.collection_select :congregation_id, Congregation.all, :id, :name, :include_blank => true
```
It's considered bad practice to make database queries in your templates. Better to have the controller function load the congregations into an attribute `@congregations = Congregation.all` and then use `@congregations` in the template.

ajdlinux replied:
ACK

tessereth commented on music/app/controllers/services_controller.rb:
```diff
@@ -11,7 +11,9 @@ def new
   def create
     @resource = OpenStruct.new(date: Date.parse(params[:resource][:date]), body: params[:resource][:body])
```
So if creating the object fails, the `new` template is re-rendered using this `@resource` to prefill the form. Normally this would be done with a model object but because this form doesn't map straight to a model, I'm faking it with OpenStruct. The provided `:congregation_id` should really be included in this command and you should check that the form remembers the value if it errors.

PS. Was going to suggest you add a test for that but... I have no testing framework for this project. So I'll take your word for it.

PPS. If for some unfathomable reason you want to add tests, use rspec, not the builtin test framework

ajdlinux replied:
"If for some unfathomable reason you want to..."

Those are daaaangerous words...

ajdlinux replied:
I've added some tests that don't quite test what you asked for, but they get part of the way...

ajdlinux committed:
Add rspec and initial ServicesController tests

Add the rspec testing framework and create some initial tests for
ServicesController to test the behaviour of the new and create actions.

Suggested-by: Tessa Bradbury <tessereth@yahoo.com.au>
Signed-off-by: Andrew Donnellan <andrew@donnellan.id.au>

ajdlinux commented:
You have internet now. ;)

tessereth commented on music/app/controllers/services_controller.rb:
```diff
@@ -1,23 +1,28 @@
     if params[:last]
       @last = Service.find_by(date: Date.parse(params[:last]))
-      @resource = OpenStruct.new(date: @last.date + 7, body: "")
+      @resource = OpenStruct.new(date: @last.date + 7, body: "", congregation_id: params[:congregation_id])
```
should this be `params[:resource][:congregation_id]`? Similarly in other places in this file.

tessereth reviewed (approved):
If you fix that one point, this LGTM.
